src/hotspot/cpu/x86/methodHandles_x86.cpp

```diff
@@ -555,18 +555,19 @@ void trace_method_handle_stub(const char* adaptername,
       PreserveExceptionMark pem(Thread::current());
       FrameValues values;
 
-      // Current C frame
-      frame cur_frame = os::current_frame();
+      frame cur_frame = os::current_frame(); // reports caller's frame!
 
       if (cur_frame.fp() != 0) {  // not walkable
 
         // Robust search of trace_calling_frame (independent of inlining).
         // Assumes saved_regs comes from a pusha in the trace_calling_frame.
         assert(cur_frame.sp() < saved_regs, "registers not saved on stack ?");
-        frame trace_calling_frame = os::get_sender_for_C_frame(&cur_frame);
+        frame trace_calling_frame = cur_frame;
```

Comment:
Since I'm expecting `cur_frame` to be `trace_method_handle_stub`,
the original code makes sense to me since I want the sender for
`trace_method_handle_stub`.

If `cur_frame` is not `trace_method_handle_stub`, then I definitely
don't understand the code at L558 so I definitely don't understand
the comment that you added either.

Reply:
No, unfortunately, it's not `trace_method_handle_stub`. It has `trace_method_handle_stub` call in tail position, so even when inlining doesn't happen, the call may be turned into a tail-call. And that's what actually happens after 8264358. So, there's no stack frame for `trace_method_handle_stub` left when `os::current_frame()` is performed. 

The next frame under top frame corresponds to the caller of `trace_method_handle_stub` in generated code (which contains registers and has to be printed). And it's the one which is returned by `os::current_frame()`.